src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/CreatePageCacheEntryActionInvoker.cs

```diff
@@ -0,0 +1,178 @@
+    /// to compile the Razor Page and cache the information required for page invocation.
+    /// </summary>
+    internal class CreatePageCacheEntryActionInvoker : IActionInvoker
+    {
+        private readonly PageLoaderBase _loader;
+        private readonly IPageFactoryProvider _pageFactoryProvider;
+        private readonly IPageModelFactoryProvider _modelFactoryProvider;
+        private readonly IModelBinderFactory _modelBinderFactory;
+        private readonly IFilterProvider[] _filterProviders;
+        private readonly IRazorPageFactoryProvider _razorPageFactoryProvider;
+        private readonly ParameterBinder _parameterBinder;
+        private readonly IModelMetadataProvider _modelMetadataProvider;
+        private readonly IPageActionInvokerFactory _pageActionInvokerFactory;
+        private readonly ConcurrentDictionary<ActionDescriptor, PageActionInvokerCacheEntry> _cache;
+
+        private readonly PageActionDescriptor _actionDescriptor;
+        private readonly ActionContext _actionContext;
+
+        public CreatePageCacheEntryActionInvoker(
+            PageLoaderBase loader,
+            IPageFactoryProvider pageFactoryProvider,
+            IPageModelFactoryProvider modelFactoryProvider,
+            IRazorPageFactoryProvider razorPageFactoryProvider,
+            IFilterProvider[] filterProviders,
+            ParameterBinder parameterBinder,
+            IModelBinderFactory modelBinderFactory,
+            IModelMetadataProvider modelMetadataProvider,
+            IPageActionInvokerFactory pageActionInvokerFactory,
+            ConcurrentDictionary<ActionDescriptor, PageActionInvokerCacheEntry> cache,
+            ActionContext actionContext,
+            PageActionDescriptor actionDescriptor)
+        {
+            _loader = loader;
+            _pageFactoryProvider = pageFactoryProvider;
+            _modelFactoryProvider = modelFactoryProvider;
+            _razorPageFactoryProvider = razorPageFactoryProvider;
+
+            _filterProviders = filterProviders;
+            _parameterBinder = parameterBinder;
+            _modelBinderFactory = modelBinderFactory;
+            _modelMetadataProvider = modelMetadataProvider;
+            _pageActionInvokerFactory = pageActionInvokerFactory;
+
+            _cache = cache;
+
+            _actionContext = actionContext;
+            _actionDescriptor = actionDescriptor;
+        }
+
+        public async Task InvokeAsync()
+        {
+            var compiledPageActionDescriptor = await _loader.LoadAsync(_actionDescriptor);
+            _actionContext.ActionDescriptor = compiledPageActionDescriptor;
+
+            var filterFactoryResult = FilterFactory.GetAllFilters(_filterProviders, _actionContext);
+            var filters = filterFactoryResult.Filters;
+            var cacheEntry = CreateCacheEntry(compiledPageActionDescriptor, filterFactoryResult.CacheableFilters);
+            cacheEntry = _cache.GetOrAdd(compiledPageActionDescriptor, cacheEntry);
```

Comment:
Does someone else read this cache? It seems like we're doing a lot of work before accessing it.